SEP-41: Add a token contract interface proposal (#1402)
stellar/stellar-protocol: ecosystem/sep-0041.md

behaviors of Stellar Assets.

An interface is needed that is a subset of the Stellar Asset contract, such that
the Stellar Asset contract would implement the interface. Tokens that implement
the interface and the Stellar Asset contract would be usable with contracts that
support the standard interface, and would be largely indistinguishable.

[CAP-46-6]: ../core/CAP-0046-06.md

## Specification

### Interface

```rust
pub trait TokenInterface {
    /// Returns the allowance for `spender` to transfer from `from`.
    ///
    /// # Arguments
    ///
    /// - `from` - The address holding the balance of tokens to be drawn from.
    /// - `spender` - The address spending the tokens held by `from`.
    fn allowance(env: Env, from: Address, spender: Address) -> i128;

    /// Set the allowance by `amount` for `spender` to transfer/burn from
    /// `from`.
    ///
    /// # Arguments
    ///
    /// - `from` - The address holding the balance of tokens to be drawn from.
    /// - `spender` - The address being authorized to spend the tokens held by
    /// `from`.
    /// - `amount` - The tokens to be made available to `spender`.

>     /// - `expiration_ledger` - The ledger number where this allowance expires.

Reviewer: ```suggestion
    /// - `live_until_ledger` - The ledger number where this allowance expires.
```

    /// Cannot be less than the current ledger number unless the amount is being

>     /// set to 0.  An expired entry (where expiration_ledger < the current

Reviewer: ```suggestion
    /// set to 0.  An expired entry (where live_until_ledger < the current
```

    /// ledger number) should be treated as a 0 amount allowance.
    ///
    /// # Events
    ///
    /// Emits an event with topics `["approve", from: Address,

>     /// spender: Address], data = [amount: i128, expiration_ledger: u32]`

Reviewer: ```suggestion
    /// spender: Address], data = [amount: i128, live_until_ledger: u32]`
```

    ///
    /// Emits an event with:
    /// - topics - `["approve", from: Address, spender: Address]`

>     /// - data - `[amount: i128, expiration_ledger: u32]`

Reviewer: ```suggestion
    /// - data - `[amount: i128, live_until_ledger: u32]`
```

>     fn approve(env: Env, from: Address, spender: Address, amount: i128, expiration_ledger: u32);

Reviewer: `expiration_ledger` is being renamed to `live_until_ledger`.

Author: ```suggestion
    fn approve(env: Env, from: Address, spender: Address, amount: i128, live_until_ledger: u32);
```

    /// Returns the balance of `id`.
    ///
    /// # Arguments
    ///
    /// - `id` - The address for which a balance is being queried. If the
    /// address has no existing balance, returns 0.
    fn balance(env: Env, id: Address) -> i128;

    /// Returns the spendable balance of `id`.
    ///
    /// If the token implementation makes a portion of the balance unavailable
    /// for general use, this should return the portion that is available for
    /// withdrawal.
    ///
    /// Should always return a value that is less than or equal to the amount
    /// returned by `balance`.
    ///
    /// # Arguments
    ///
    /// - `id` - The address for which a spendable balance is being queried.

>     fn spendable_balance(env: Env, id: Address) -> i128;

Reviewer: Does this interface need a `spendable_balance` (I think this was discussed recently but don't remember what came out of that).

Author: The discussion and resolution is here:
- https://github.com/stellar/rs-soroban-sdk/issues/1080#issuecomment-1707402077

The resolution was we could go both ways so leave it. But, I've lost count how many times this has come up, and every time the opinion is "it doesn't belong."

I think we could remove it with little impact.

cc @tomerweller
